source/base/tensor_function_parser.cc

```diff
@@ -0,0 +1,390 @@
+
+  // finally set the initialization bit
+  initialized = true;
+}
+
+
+
+template <int rank, int dim, typename Number>
+void
+TensorFunctionParser<rank, dim, Number>::init_muparser() const
+{
+  // check that we have not already initialized the parser on the
+  // current thread, i.e., that the current function is only called
+  // once per thread
+  Assert(tfp.get().size() == 0, ExcInternalError());
+
+  // initialize the objects for the current thread (tfp.get() and
+  // vars.get())
+  tfp.get().reserve(this->n_components);
+  vars.get().resize(var_names.size());
+  for (unsigned int component = 0; component < this->n_components; ++component)
+    {
+      tfp.get().emplace_back(new mu::Parser());
+
+      for (std::map<std::string, double>::const_iterator constant =
+             constants.begin();
+           constant != constants.end();
+           ++constant)
+        {
+          tfp.get()[component]->DefineConst(constant->first, constant->second);
+        }
+
+      for (unsigned int iv = 0; iv < var_names.size(); ++iv)
+        tfp.get()[component]->DefineVar(var_names[iv], &vars.get()[iv]);
+
+      // define some compatibility functions:
+      tfp.get()[component]->DefineFun("if", internal::mu_if, true);
+      tfp.get()[component]->DefineOprt("|", internal::mu_or, 1);
+      tfp.get()[component]->DefineOprt("&", internal::mu_and, 2);
+      tfp.get()[component]->DefineFun("int", internal::mu_int, true);
+      tfp.get()[component]->DefineFun("ceil", internal::mu_ceil, true);
+      tfp.get()[component]->DefineFun("cot", internal::mu_cot, true);
+      tfp.get()[component]->DefineFun("csc", internal::mu_csc, true);
+      tfp.get()[component]->DefineFun("floor", internal::mu_floor, true);
+      tfp.get()[component]->DefineFun("sec", internal::mu_sec, true);
+      tfp.get()[component]->DefineFun("log", internal::mu_log, true);
+      tfp.get()[component]->DefineFun("pow", internal::mu_pow, true);
+      tfp.get()[component]->DefineFun("erfc", internal::mu_erfc, true);
+      tfp.get()[component]->DefineFun("rand_seed",
+                                      internal::mu_rand_seed,
+                                      true);
+      tfp.get()[component]->DefineFun("rand", internal::mu_rand, true);
+
+      try
+        {
+          // muparser expects that functions have no
+          // space between the name of the function and the opening
+          // parenthesis. this is awkward because it is not backward
+          // compatible to the library we used to use before muparser
+          // (the tfparser library) but also makes no real sense.
+          // consequently, in the expressions we set, remove any space
+          // we may find after function names
+          std::string transformed_expression = expressions[component];
+
+          internal::MuParserFunctionNames function_names;
+          for (auto function_name_it = function_names.begin();
+			  function_name_it != function_names.end();
+			  ++function_name_it)
```

Comment:
This also looks like a place a range-based for loop can be used.

Reply:
OK, I am using a range based for loop now.